addons/binding/org.openhab.binding.nibeheatpump/src/main/java/org/openhab/binding/nibeheatpump/internal/connection/SerialConnector.java

```diff
@@ -0,0 +1,371 @@
+            return;
+        }
+
+        conf = configuration;
+        try {
+            CommPortIdentifier portIdentifier = CommPortIdentifier.getPortIdentifier(conf.serialPort);
+
+            CommPort commPort = portIdentifier.open(this.getClass().getName(), 2000);
+
+            serialPort = (SerialPort) commPort;
+            serialPort.setSerialPortParams(9600, SerialPort.DATABITS_8, SerialPort.STOPBITS_1, SerialPort.PARITY_NONE);
+
+            serialPort.enableReceiveThreshold(1);
+            serialPort.disableReceiveTimeout();
+
+            in = serialPort.getInputStream();
+            out = serialPort.getOutputStream();
+
+            out.flush();
+            if (in.markSupported()) {
+                in.reset();
+            }
+        } catch (NoSuchPortException e) {
+            throw new NibeHeatPumpException("Connection failed, no such a port", e);
+        } catch (PortInUseException e) {
+            throw new NibeHeatPumpException("Connection failed, port in use", e);
+        } catch (UnsupportedCommOperationException | IOException e) {
+            throw new NibeHeatPumpException("Connection failed, reason: {}" + e.getMessage(), e);
+        }
+
+        readQueue.clear();
+        writeQueue.clear();
+
+        readerThread = new SerialReader(in);
+        readerThread.start();
+        connected = true;
+    }
+
+    @Override
+    public void disconnect() throws NibeHeatPumpException {
+
+        logger.debug("Disconnecting");
+
+        serialPort.removeEventListener();
+
+        if (readerThread != null) {
+            logger.debug("Interrupt serial listener");
+            readerThread.interrupt();
+        }
+
+        if (out != null) {
+            logger.debug("Close serial out stream");
+            IOUtils.closeQuietly(out);
+        }
+        if (in != null) {
+            logger.debug("Close serial in stream");
+            IOUtils.closeQuietly(in);
+        }
+
+        if (serialPort != null) {
+            logger.debug("Close serial port");
+            serialPort.close();
+        }
+
+        readerThread = null;
+        serialPort = null;
+        out = null;
+        in = null;
+        connected = false;
+        logger.debug("Closed");
+    }
+
+    @Override
+    public void sendDatagram(NibeHeatPumpMessage msg) throws NibeHeatPumpException {
+        logger.trace("Add request to queue: {}", msg.toHexString());
+
+        if (msg instanceof ModbusWriteRequestMessage) {
+            writeQueue.add(msg.decodeMessage());
+        } else if (msg instanceof ModbusReadRequestMessage) {
+            readQueue.add(msg.decodeMessage());
+        } else {
+            logger.trace("Ignore PDU: {}", msg.getClass().toString());
+        }
+
+        logger.trace("Read queue: {}, Write queue: {}", readQueue.size(), writeQueue.size());
+    }
+
+    public class SerialReader extends Thread implements SerialPortEventListener {
+        boolean interrupted = false;
+        InputStream in;
```

Comment:
One reason is the RXTX library CPU load workaround (see line 315 ->). Not sure how it works is it's runnable rather than thread.